File: LeetCodePython/py59.py
# 螺旋矩阵 II
# 填充1~n*b到按顺时针顺序螺旋排列正方形矩阵
from typing import List
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        left = 0
        right = n-1
        up = 0
        down = n-1
        result = [[0]*n for _ in range(n)]
        num = 1
        while True:
            for i in range(left, right+1):  # 往右走
                result[up][i] = num
                num += 1
            up += 1
            
            if up > down:  
                break
            for i in range(up, down+1):  # 往下走
                result[i][right] = num
                num += 1
            
            right -= 1
            if right<left:
                break
            
            for i in range(right, left-1, -1): # 往左走
                result[down][i] = num
                num += 1
            
            down -= 1
            if down < up:
                break
      
            for i in range(down, up-1, -1): # 往上走
                result[i][left] = num
                num += 1
      
            left += 1
            if left>right:
                break
        return result

File: LeetCodeCPlusPlus/py59.cpp
// 螺旋矩阵II
#include <vector>
using namespace std;
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> result(n, vector<int>(n));
        int left = 0, right = n - 1, up = 0, down = n - 1, num = 1;
        while (true) {
            for (int i=left; i<=right; i++)  //往右
                result[up][i] = num++;
            if (++up>down) break;
            for (int i=up; i<=down; i++) // 往下
                result[i][right] = num++;
            if (--right<left) break;
            for (int i=right; i>=left; i--) // 往左
                result[down][i] = num++;
            if (--down<up) break;
            for (int i=down; i>=up; i--) // 往上走
                result[i][left] = num++;
            if (++left>right) break;
        }
        return result;
    }
};